File: modules/video-pose-analysis/android/src/main/java/com/divotgolf/videoposeanalysis/VideoPosePipeline.kt
package com.divotgolf.videoposeanalysis

import android.content.Context
import android.util.Log

/**
 * Orchestrates frame extraction → pose detection across all frames of a video.
 * Mirrors the iOS VideoPosePipeline structure.
 */
class VideoPosePipeline(
    private val context: Context,
    private val filePath: String,
) {
    companion object {
        private const val TAG = "VideoPoseAnalysis"
        private const val PROGRESS_INTERVAL = 10
    }

    /** Called every N frames with progress (0-1), currentFrame, totalFrames. */
    var onProgress: ((Double, Int, Int) -> Unit)? = null

    /** Set to true to cancel processing between frames. */
    @Volatile
    var isCancelled = false

    /**
     * Runs pose detection on every frame and returns results as a map
     * matching the TypeScript VideoPoseAnalysisResult type.
     */
    fun run(clipId: String): Map<String, Any> {
        val startTime = System.currentTimeMillis()
        val detector = VideoPoseDetector(context)

        if (!detector.isReady) {
            throw IllegalStateException("Pose detection model failed to load")
        }

        val extractor = VideoFrameExtractor(filePath)
        val totalFrames = extractor.totalFrames
        val fps = extractor.frameRate
        val rotationDegrees = extractor.rotationDegrees

        Log.d(TAG, "Starting pose pipeline: $totalFrames frames at $fps fps")

        val frameResults = mutableListOf<Map<String, Any>>()
        var analysisWidth = 0
        var analysisHeight = 0

        extractor.use { ext ->
            ext.extractFrames { frameIndex, timestampMs, bitmap ->
                if (isCancelled) {
                    Log.d(TAG, "Analysis cancelled at frame $frameIndex")
                    return@extractFrames false
                }

                if (frameIndex % PROGRESS_INTERVAL == 0) {
                    val progress = frameIndex.toDouble() / totalFrames.toDouble()
                    onProgress?.invoke(progress, frameIndex, totalFrames)
                }

                if (analysisWidth == 0) {
                    analysisWidth = bitmap.width
                    analysisHeight = bitmap.height
                    Log.d(TAG, "Actual frame size: ${analysisWidth}x$analysisHeight")
                }

                val landmarks = detector.detectPose(bitmap, timestampMs, rotationDegrees)
                if (landmarks != null) {
                    frameResults.add(mapOf(
                        "frameIndex" to frameIndex,
                        "timestampMs" to timestampMs.toDouble(),
                        "landmarks" to landmarks,
                    ))
                }

                true // continue
            }
        }

        detector.close()

        if (isCancelled) {
            throw InterruptedException("Analysis was cancelled")
        }

        // Final progress
        onProgress?.invoke(1.0, totalFrames, totalFrames)

        val analysisTimeMs = (System.currentTimeMillis() - startTime).toDouble()

        Log.d(TAG, "Pose pipeline complete: ${frameResults.size}/$totalFrames frames detected in ${analysisTimeMs.toInt()}ms")

        return mapOf(
            "clipId" to clipId,
            "totalFrames" to totalFrames,
            "analyzedFrames" to frameResults.size,
            "frames" to frameResults,
            "analysisTimeMs" to analysisTimeMs,
            "fps" to fps.toDouble(),
            "resolution" to mapOf(
                "width" to analysisWidth,
                "height" to analysisHeight,
            ),
        )
    }
}

File: modules/video-pose-analysis/ios/VideoPosePipeline.swift
import CoreMedia
import Foundation
import os.log

private let logger = Logger(subsystem: "com.divotgolf.videoposeanalysis", category: "VideoPosePipeline")

/// Orchestrates frame extraction → pose detection across all frames of a video.
/// Returns a result dictionary matching the TypeScript VideoPoseAnalysisResult type.
final class VideoPosePipeline {
    private let extractor: VideoFrameExtractor
    private let detector: VideoPoseDetector

    /// Called every N frames with progress (0-1), currentFrame, totalFrames.
    var onProgress: ((Double, Int, Int) -> Void)?

    /// Set to true to cancel processing between frames.
    var isCancelled = false

    /// How often to report progress (every N frames).
    private let progressInterval = 10

    init(url: URL) throws {
        extractor = try VideoFrameExtractor(url: url)
        detector = VideoPoseDetector()

        guard detector.isReady else {
            throw VideoPoseError.modelNotAvailable
        }
    }

    /// Runs pose detection on every frame and returns results.
    func run(clipId: String) throws -> [String: Any] {
        let startTime = CFAbsoluteTimeGetCurrent()
        let totalFrames = extractor.totalFrames
        let fps = extractor.frameRate

        var imageWidth = 0
        var imageHeight = 0

        logger.info("Starting pose pipeline: \(totalFrames) frames at \(fps) fps")

        var frameResults: [[String: Any]] = []

        for frameIndex in 0..<totalFrames {
            if isCancelled {
                logger.info("Analysis cancelled at frame \(frameIndex)")
                throw VideoPoseError.cancelled
            }

            if frameIndex % progressInterval == 0 {
                let progress = Double(frameIndex) / Double(totalFrames)
                onProgress?(progress, frameIndex, totalFrames)
            }

            guard let (cgImage, actualTime) = extractor.extractFrame(at: frameIndex) else {
                continue
            }

            if imageWidth == 0 {
                imageWidth = cgImage.width
                imageHeight = cgImage.height
                logger.info("Actual frame size: \(imageWidth)x\(imageHeight)")
            }

            let timestampMs = CMTimeGetSeconds(actualTime) * 1000.0

            guard let landmarks = detector.detectPose(in: cgImage) else {
                continue
            }

            let frameResult: [String: Any] = [
                "frameIndex": frameIndex,
                "timestampMs": timestampMs,
                "landmarks": landmarks,
            ]

            frameResults.append(frameResult)
        }

        // Final progress
        onProgress?(1.0, totalFrames, totalFrames)

        let analysisTimeMs = (CFAbsoluteTimeGetCurrent() - startTime) * 1000.0

        logger.info("Pose pipeline complete: \(frameResults.count)/\(totalFrames) frames detected in \(Int(analysisTimeMs))ms")

        return [
            "clipId": clipId,
            "totalFrames": totalFrames,
            "analyzedFrames": frameResults.count,
            "frames": frameResults,
            "analysisTimeMs": analysisTimeMs,
            "fps": fps,
            "resolution": [
                "width": imageWidth,
                "height": imageHeight,
            ],
        ]
    }
}
